Makefile.docker

```diff
@@ -0,0 +1,95 @@
+
+DOCKER_RUN := docker run --rm -w /build -v $(shell pwd):/build:Z $(IMAGE)
+
+# This is to reliably maintain permissions in a cross compatible manner
+# (tested on Linux and Mac). The docker container needs the correct
+# permissions to build artifcats, but also needs to give ownership of build
+# artifacts back to the host machine.
+CHOWN := chown -R $(shell id -u):$(shell id -g) .
+
+ifeq ($(ALL), 1)
+	CLEAN_ADDTL_FILTERS :=
+else
+	CLEAN_ADDTL_FILTERS := --filter label=com.logdna.tag=$(BUILD_TAG)
+endif
+
+CLEAN_DOCKER_IMAGES = if [[ ! -z "docker images -q --filter $(1) $(CLEAN_ADDTL_FILTERS)" ]]; then docker images -q --filter $(1) $(CLEAN_ADDTL_FILTERS) | head -n 1 | xargs -I {} docker rmi -f {}; fi
+
+.PHONY:build
+build: ## (Runs in a docker container) Build the agent. Set RELEASE=1 to build a release image - defaults to 0
+	$(DOCKER_RUN) /bin/sh -c "make build RELEASE=$(RELEASE) && $(CHOWN)"
+
+.PHONY:test
+test: ## (Runs in a docker container) Installs the necessary dependencies and then checks rust syntax, docker syntax, rust warnings, outdated dependencies, security vulnerabilities, and unit tests
+	$(DOCKER_RUN) /bin/sh -c "make test-deps && make test && $(CHOWN)"
+
+.PHONY:clean
+clean: ## (Runs in a docker container) Clean all artifacts from the build process
+	$(DOCKER_RUN) /bin/sh -c "make clean && $(CHOWN)"
+
+.PHONY:clean-images
+clean-images: ## Cleans the intermediate and final agent images left over from the build-image target
+	@# Clean any intermediate agent images, left over from the multi-stage build
+	$(call CLEAN_DOCKER_IMAGES,label=com.logdna.stage=$(REPO)-build)
+	@# Clean any final agent images
+	$(call CLEAN_DOCKER_IMAGES,label=com.logdna.stage=$(REPO)-final)
+
+.PHONY:build-image
+build-image: ## Build a docker image as specified in the Dockerfile
+	$(DOCKER) build . -t $(REPO):$(BUILD_TAG) \
+		--pull --no-cache=true --rm \
+		--build-arg BUILD_IMAGE=$(IMAGE) \
+		--build-arg BUILD_TIMESTAMP=$(BUILD_TIMESTAMP) \
+		--build-arg BUILD_VERSION=$(BUILD_VERSION) \
+		--build-arg REPO=$(REPO) \
+		--build-arg VCS_REF=$(VCS_REF) \
+		--build-arg VCS_URL=$(VCS_URL)
+	$(DOCKER) tag $(REPO):$(BUILD_TAG) $(REPO):$(BUILD_VERSION)
+
+.PHONY:publish-private
+publish-private: ## Publish a build version of the docker image to our private registry
+	for version in $(BUILD_TAG) $(BUILD_VERSION); do \
+		$(DOCKER) tag $(REPO):$(BUILD_TAG) $(DOCKER_PRIVATE_IMAGE):$${version}; \
+		$(DOCKER) push $(DOCKER_PRIVATE_IMAGE):$${version}; \
+	done;
+
+.PHONY:publish-public
+publish-public: ## Publish SemVer compliant releases to our public registries
+	@# TODO: Have a boolean that prevents this unless forced or run by Jenkins (which can force it)
+	for image in $(DOCKER_PUBLIC_IMAGE) $(DOCKER_IBM_IMAGE); do \
+		for version in $(MAJOR_VERSION) $(MINOR_VERSION) $(PATCH_VERSION) $(BUILD_VERSION) latest; do \
+			if [[ $${version} != $(BUILD_VERSION) ]] && [[ $(PRERELEASE_VERSION) = beta* ]]; then \
+				continue; \
+			fi; \
```

Comment:
If the version is a beta, we won't update the major, minor, or patch versions of the images. E.g. for `2.1.9-beta.2`, the only thing published to the public is the tag `2.1.9-beta.2`

Reply:
You shouldnt need to buiuld i nthe extra logic. the pipeline in jenkins should be configured to push it only if its being promoted.  such as [here](https://github.com/answerbook/tooling-elasticsearch/blob/master/Jenkinsfile#L17-L38)